iroh-bytes/src/downloader/test.rs

```diff
@@ -345,3 +361,134 @@ async fn long_queue() {
         res.expect("all downloads to succeed");
     }
 }
+
+#[tokio::test]
+async fn fail_while_running() {
```

Comment:
Your intuition was correct. The branch previously had a bug (spotted by @ppodolsky) where `DropPeer` would not first check if there are other downloads running but drop the peer unconditionally. The test recreates that scenario, and passes since the bug was fixed.
I added this comment to the test:
```rust
/// If a download errors with [`FailureAction::DropPeer`], make sure that the peer is not dropped
/// while other transfers are still running.
```

>  Does it matter whether the peer was or not removed (DropPeer) after both are done?

Currently, a peer will only be dropped after a `FailureAction::DropPeer` if no other transfer is running. My intuition for that choice: 
* If the other transfers after the failed one succeed without error, the condition that caused DropPeer was likely transient, so we just keep the peer and ignore the previous DropPeer
* If the failure condition remains, the other transfers will fail with DropPeer as well, and after the last transfer failed, the peer will be dropped.

As I'm typing this, the following scenario would be possible in the current implementation, but likely should be avoided:
* Queue hashA, hashB with nodeA. hashA, hashB are started
* hashA fails with DropPeer -> peer is not dropped because hashB is still running
* Queue hashC with nodeA. hashC is started
* hashB fails with DropPeer -> peer is not dropped  because hashC is still running
* Queue hashD with nodeA. hashD is started
* ... repeat as above

In plain words: If I keep queuing new dowloads to a bad node in some frequency, it might happen that the peer is never dropped even though it is clearly faulty. Downloads will still succeed if other providers are available, but cycles will be waisted and resource capacity needlessly consumed.

I think I will add a test for this case. A fix could be to mark a node as `failed` (or count the failures) and not queue new downloads for suchly marked nodes (but do not drop them while things are running). After a successful transfer, the `failed` marker is cleared.